src/Main.hs

```diff
@@ -74,11 +105,40 @@ proceed' el = do
 untwistDeps :: [Element] -> DatatypeMap -> ([(Text, Datatype)], [Element])
 untwistDeps e dm = (dList, e')
   where
-    dList   = nub $ execWriter $ do
-      (a, (_,s)) <- runStateT sortDatatypesByDeps (dm, S.empty)
+    dList   = deconflictNames . nub . execWriter $ do
+      (a, _) <- runStateT sortDatatypesByDeps (dm, S.empty)
       return a
     (_, e') = execRWS sortElemsByDeps (M.fromList dList) e
 
+type Abbreviation = Text
+type Subelement = Text
+type DeconflicterMonad = State (M.Map Abbreviation (S.Set Subelement))
+
+deconflictNames :: [(Text, Datatype)] -> [(Text, Datatype)]
+deconflictNames = flip evalState mempty . mapM deconflictName
+
+deconflictName :: (Text, Datatype) -> DeconflicterMonad (Text, Datatype)
```

Comment:
Probably should be called `disambiguate` instead. Please add a haddock describing the problem and solution here.